sdk/storage/azure-storage-datamovement/inc/azure/storage/datamovement/datamovement_options.hpp

```diff
@@ -3,22 +3,64 @@
 
 #pragma once
 
+#include <functional>
+#include <memory>
 #include <string>
 
+#include <azure/core/credentials/credentials.hpp>
+#include <azure/storage/common/storage_credential.hpp>
+
 namespace Azure { namespace Storage {
   struct StorageTransferManagerOptions final
   {
     std::string TransferStateDirectoryPath;
+    Nullable<int> NumThreads;
+    Nullable<size_t> MaxMemorySize;
   };
 
-  namespace Blobs {
+  struct TransferCredential final
+  {
+    std::string ConnectionString;
+    std::string SasCredential; // TODO: or Azure SAS credential?
+    std::shared_ptr<StorageSharedKeyCredential> SharedKeyCredential;
+    std::shared_ptr<Core::Credentials::TokenCredential> TokenCredential;
+  };
+
+  struct TransferProgress final
+  {
+    int64_t NumFilesTransferred = 0;
+    int64_t NumFilesSkipped = 0;
+    int64_t NumFilesFailed = 0;
+    int64_t TotalBytesTransferred = 0;
+  };
 
+  struct TransferError final
+  {
+    std::string JobId;
+    std::string SourceUrl;
+    std::string DestinationUrl;
+    // TODO: error information: std::exception_ptr or descriptive string?
```

Comment:
An exception instance should be better. We may also need to report to customer about details of requests.

Reply:
I'm thinking the opposite. As you may have noticed, task must not throw exceptions when it's running. Considering the fact that error handler is executed synchronously in tasks and exception will be triggered there when customer tries to observe the exception ptr, I'm worried that the whole scheduling system will be messed up if they don't properly handle the exception.

If we use exception ptr here, we may need another layer of security to ensure exception won't escape from task scope. 
Let's keep the TODO here and revisit this later.

Reply:
We can catch all exception thrown during execution and report them to customer through the callback. Scheduler doesn't need to know anything about this. Anyway, let's a TODO here for now.